chainer/dataset/convert.py

```diff
@@ -139,3 +140,181 @@ def _concat_arrays_with_padding(arrays, padding):
+        >>>               ...)
+
+    Args:
+        stream (cupy.cuda.Stream): CUDA stream. If ``None``, a stream is
+            automatically created on the first call. Data transfer operation
+            is launched acynchrnously using the stream.
+    """
+
+    def __init__(self, stream=None):
+        self._stream = stream
+        self._device = None
+        self._conveyor = collections.defaultdict(
+            lambda: Conveyor(self._device, self._stream))
+
+    def __call__(self, batch, device=None, padding=None):
+        """Concatenate data and transfer them to GPU asynchronously.
+
+        (see concat_examples for detail about concatenation)
+
+        Args:
+            batch (list): A list of examples.
+            device (int): Device ID to which each array is sent.
+            padding: Scalar value for extra elements.
+
+        Returns:
+            Array, a tuple of arrays, or a dictionary of arrays.
+            The type depends on the type of each example in the batch.
+        """
+        if len(batch) == 0:
+            raise ValueError('batch is empty')
+        first_elem = batch[0]
+
+        if len(self._conveyor) == 0:
+            self._device = device  # device is set at first call
+            if device is not None and device >= 0 and self._stream is None:
+                with cuda.get_device_from_id(device):
+                    self._stream = cuda.Stream(non_blocking=True)
+        if device is not self._device:
+            raise ValueError('device is different')
+
+        with cuda.get_device_from_id(device):
+            if isinstance(first_elem, tuple):
+                result = []
+                if not isinstance(padding, tuple):
+                    padding = [padding] * len(first_elem)
+
+                for i in six.moves.range(len(first_elem)):
+                    self._conveyor[i].put(_concat_arrays(
+                        [example[i] for example in batch], padding[i]))
+
+                for i in six.moves.range(len(first_elem)):
+                    result.append(self._conveyor[i].get())
+
+                return tuple(result)
+
+            elif isinstance(first_elem, dict):
+                result = {}
+                if not isinstance(padding, dict):
+                    padding = {key: padding for key in first_elem}
+
+                for key in first_elem:
+                    self._conveyor[key].put(_concat_arrays(
+                        [example[key] for example in batch], padding[key]))
+
+                for key in first_elem:
+                    result[key] = self._conveyor[key].get()
+
+                return result
+
+            else:
+                return to_device(device, _concat_arrays(batch, padding))
+
+
+class Conveyor(object):
+
+    """Interface to handle asynchronous data transfer using double buffering.
+
+    An asynchrous data transfer is initiated by :meth:`put`, and the result,
+    the array transferd to a target device, is obtained by :meth:`get`.
```

Comment:
transferred?

Reply:
Thank you for pointing it out! will fix it.